### DISCUSSION TITLE
Hard-burn subtitles

### DISCUSSION THREAD

**Discussion by philstopford** [CineEncoder/cine-encoder#68] [Q&A]
So I've been hacking away at this for a while and seem to have run into an odd wall that I don't yet understand. In the PRs I've submitted, it looks like it should be working, but I get an error reported. Printing out the arguments, I can copy and paste them with the installed ffmpeg and the hard-burn works without issue or error:

```
-hide_banner -probesize 100M -analyzeduration 50M -i /home/phil/Downloads/my.test.file.mp4 -map 0:v:0? -map 0:a:0?  -vf "subtitles=/home/phil/Downloads/my.test.file.srt" -map_metadata -1 -map_chapters -1 -map_metadata:s:v:0 -1 -metadata:s:a:0 language= -metadata:s:a:0 title= -disposition:a:0 default  -pix_fmt yuv420p -c:v libx265 -profile:v main -preset fast -crf 25 -c:a aac -b:a 256k -y /home/phil/Downloads/my.test.file_encoded_0000.mp4
```

I can't see the exact nature of the error so far - this is what gets reported from cineencoder itself:

```
An error occurred:  Stream #0:1[0x2](und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 127 kb/s (default)
 Metadata:
 handler_name : SoundHandler
 vendor_id : [0][0][0][0]
[NULL @ 0x555555716440] Unable to find a suitable output format for ''
: Invalid argument
```

I'm baffled.

![image](https://user-images.githubusercontent.com/1983851/235569310-268e13d8-ef31-4c70-9814-dc607d8ccd54.png)

The first snippet above is what I get for args_ and runs just fine, as mentioned, if I copy/paste into a terminal and run with ffmpeg.

Any ideas?

**Comment by CineEncoder** [comment]
Hello! 
So far it is really not clear, it will be necessary to study this problem.

**Comment by philstopford** [comment]
Just wondering if you'd made any progress. I remain confused, sadly.

**Comment by CineEncoder** [comment]
Hello! Still no idea why it doesn't work.

**Comment by philstopford** [comment]
I'm still at a loss - tried everything I can think of to understand what's misfiring, no joy. Really weird.

**Comment by philstopford** [comment]
@CineEncoder : did you manage to figure this out? I'm still puzzled, sadly.

**Reply by CineEncoder** [reply]
Hello!
I was able to do this for the case where the filename does not contain spaces or quotes, there was some strange problem with escape sequences when using the vf filter:
https://github.com/CineEncoder/cine-encoder/blob/prototype/hardburn_subtitle_r1/app/encoder.cpp#L280
When using single quotes, a filename containing spaces is not treated as a single line, and when using the escape sequence `\" \"`, an error occurs. Also using replace spaces with `"\\ "` does not give any result.

I slightly modified the code, made my version on the new branch in one commit:
https://github.com/CineEncoder/cine-encoder/commit/7ccfc8d3fa421d9b35e94cdaf8d9310e550cc5a5
There is also a problem when using splitting and this option at the same time: the subtitles are not synchronized in time in this case, as far as I understand.

**Reply by philstopford** [reply]
Outstanding, thank you! I'm not quite sure what I missed in the original code to cause the problem originally. I was thinking about options to configure font, size and whether or not to have a background behind the subtitle burn-in, to improve readability. Given the UI design, I'm tempted to think that this would be a set of global preferences used when burning in subtitles.

I wanted to ask before I started poking at the code in case you preferred to tackle that rather than get an unwelcome / unhelpful PR from me :)

**Reply by CineEncoder** [reply]
Hello! Sorry for the delay. Yes, I think the subtitle font settings can be added in the preset settings, between the "Audio" and "Metadata" tabs it will be possible to create a "Subtitle" tab.
![subtitle](https://github.com/CineEncoder/cine-encoder/assets/68183463/86fd2c38-d55f-438f-9daf-ca1527dc1945)
So far, it is still not clear what to do with spaces in file names for embedded subtitles.

**Reply by philstopford** [reply]
For the subtitles file, it might be enough to make a copy that has a legal name that we can then pick up and use, e.g. hash the existing filename, make a copy to that temporary name in (temp) and then use that.

For the UI, I was considering something under Preferences so that the user will get the same font settings (including background) when enabling hard-burning, without having to worry about presets. I don't have a strong view either way.

**Comment by philstopford** [comment]
So I've submitted a PR here : https://github.com/CineEncoder/cine-encoder/pull/70

This re-works the internals to avoid constructing and then decomposing strings. The old approach was not particularly robust:
 - it assumed that you could always split arguments and values by space, to compose the arguments list, but this breaks when any input contains spaces.
 - you had to pay attention to any duplicated spaces (e.g. "   ") because split would yield multiple empty entries in the arguments lists, which also broke the ffmpeg invocation.

Moving to QStringLists gives really convenient separation automatically, and it also allows empty entries to be easily cleaned out.

I've also added UI controls for hard-burn selection of font, color, whether to add a background, background color and background opacity.

I've tested most of this and it seems to work, but more testing is welcome.

**Comment by philstopford** [comment]
Location is also now added, and testing revealed and allowed fixing of various issues. I'm fairly happy with this. I'm tempted to look at the preset system to see if there might be a way to have preset-defined subtitles that could override the preferences definitions, so that both workflows can be accommodated.

**Comment by philstopford** [comment]
I set up a WIP branch for preset implementation. I am struggling to see how the preset UI is updated based on the preset values at the moment : https://github.com/philstopford/cine-encoder/tree/hardburn_subtitles_presetWIP

**Comment by philstopford** [comment]
@CineEncoder : some overview / assistance for the preset side of things would be welcome. I'm actually wondering about the design of this system because it seems to need a rebuild each time the preset changes. I wondered whether something around XML would work better rather than what appears to be a binary blob as well.

What do you think?

**Comment by philstopford** [comment]
Just nudging re preset guidance
